packages/coreutils/src/windowresolver.ts

```diff
@@ -0,0 +1,222 @@
+ */
+export
+interface IWindowResolver {
+  /**
+   * A window name to use as a handle among shared resources.
+   */
+  readonly name: string;
+}
+
+
+/**
+ * A concrete implementation of a window name resolver.
+ */
+export
+class WindowResolver implements IWindowResolver {
+  /**
+   * The resolved window name.
+   */
+  get name(): string {
+    return this._name;
+  }
+
+  /**
+   * Resolve a window name to use as a handle among shared resources.
+   *
+   * @param candidate - The potential window name being resolved.
+   *
+   * #### Notes
+   * Typically, the name candidate should be a JupyterLab workspace name or
+   * an empty string if there is no workspace.
+   *
+   * If the returned promise rejects, a window name cannot be resolved without
+   * user intervention, which typically means navigation to a new URL.
+   */
+  resolve(candidate: string): Promise<void> {
+    return Private.resolve(candidate).then(name => { this._name = name; });
+  }
+
+  private _name: string | null = null;
+}
+
+
+/*
+ * A namespace for private module data.
+ */
+namespace Private {
+  /**
+   * The internal prefix for private local storage keys.
+   */
+  const PREFIX = '@jupyterlab/coreutils:StateDB';
+
+  /**
+   * The local storage beacon key.
+   */
+  const BEACON = `${PREFIX}:beacon`;
+
+  /**
+   * The timeout (in ms) to wait for beacon responders.
+   *
+   * #### Notes
+   * This value is a whole number between 200 and 500 in order to prevent
+   * perfect timeout collisions between multiple simultaneously opening windows
+   * that have the same URL. This is an edge case because multiple windows
+   * should not ordinarily share the same URL, but it can be contrived.
+   */
+  const TIMEOUT = Math.floor(200 + Math.random() * 300);
+
+  /**
+   * The local storage window key.
+   */
+  const WINDOW = `${PREFIX}:window`;
+
+  /**
+   * A potential preferred default window name.
+   */
+  let candidate: string | null = null;
+
+  /**
+   * The window name promise.
+   */
+  let delegate = new PromiseDelegate<string>();
+
+  /**
+   * The known window names.
+   */
+  let known: { [window: string]: null } = { };
+
+  /**
+   * The window name.
+   */
+  let name: string | null = null;
+
+  /**
+   * Whether the name resolution has completed.
+   */
+  let resolved = false;
+
+  /**
+   * Start the storage event handler.
+   */
+  function initialize(): void {
+    // If this library is not running in the browser, bail.
+    if (typeof window === 'undefined') {
+      return;
+    }
+
+    // Listen to all storage events for beacons and window names.
+    window.addEventListener('storage', (event: StorageEvent) => {
+      const { key, newValue } = event;
+
+      // All the keys we care about have values.
+      if (newValue === null) {
+        return;
+      }
+
+      // If the beacon was fired, respond with a ping.
+      if (key === BEACON && candidate !== null) {
+        return ping(resolved ? name : candidate);
```

Comment:
This feels very nautical :boat:

What is the purpose of returning the value of `ping()` here (which appears to be unconditionally `undefined`)?

Reply:
That was just shorthand, but I should adhere to our regular style and make the `return;` statement stand on its own!